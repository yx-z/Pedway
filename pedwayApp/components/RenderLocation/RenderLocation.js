import React, {Component} from 'react';
import PedwayCoordinate from '../../model/PedwayCoordinate';
import PedwayEntrance from '../../model/PedwayEntrance';
import MapView from 'react-native-maps';

/**
 * The current pedway sections are hard coded place holders
 * In the future we are gonna to get those values from the API
 * */
export default class RenderLocation extends Component {
  constructor(props) {
    super(props);
    this.state = {
      pedwayLocations: [],
    };
    this.parseJSONtoModel = this.parseJSONtoModel.bind(this);
  }

  /**
   * parses the representation of a list of locations fetched from the point of interest api
   * into a list of locations
   * the coordinate and status will be set according to the inputJSON
   * @param inputJSON
   */
  parseJSONtoModel(inputJSON) {
    const locations = inputJSON.filter((item) => {
      try {
        if (item['geometry']['type'] === 'Point') {
          return true;
        }
        return false;
      } catch (e) {
        return false;
      }
    }).reduce((acc, item) => {
      const thisLongitude = item['geometry']['coordinates'][0];
      const thisLatitude = item['geometry']['coordinates'][1];
      const thisName = item['properties']['name'];
      return acc.concat(
          new PedwayEntrance(new PedwayCoordinate(
              thisLatitude,
              thisLongitude),
          true,
          false,
          thisName));
    }, []);
    this.setState({
      pedwayLocations: locations,
    });
  }

  componentWillMount() {
    if (this.props.JSONData!==undefined) {
      this.parseJSONtoModel(this.props.JSONData);
    }
  }

  componentWillReceiveProps(next) {
    if (next.JSONData!==undefined) {
      this.parseJSONtoModel(next.JSONData);
    }
    this.forceUpdate();
  }

  /**
   * renders the location list parsed from the inputJSON into a list of markers on the map
   * onclickListener is setup for the markers so that the user can click on a marker and bring up
   * the corresponding slingupview
   * @returns {*[]}
   */
  render() {
    const retMarkerList = this.state.pedwayLocations.map((input, idx) => {
      return (
        <MapView.Marker
          coordinate={input.getCoordinate().getJSON()}
          key={idx}
          onPress={()=>{
            this.props.callbackFunc(this.state.pedwayLocations[idx], false);
          }}
        />
      );
    },
    );
    return (
      retMarkerList
    );
  }
}
