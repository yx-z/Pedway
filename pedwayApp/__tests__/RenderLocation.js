import RenderLocation from '../components/RenderLocation/RenderLocation';
import testRenderer from 'react-test-renderer';
import React from 'react';


test('Check if our RenderLocation renders correctly', () => {
  const treeRendered = testRenderer.create(<RenderLocation />).toJSON();
  expect(treeRendered).toMatchSnapshot();
});


