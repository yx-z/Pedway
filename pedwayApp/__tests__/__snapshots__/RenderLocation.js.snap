// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if our RenderLocation renders correctly 1`] = `null`;
